## Day 1

In [184]:
import pandas as pd
from scipy.io import arff

# Load the arff file
data = arff.loadarff('../data/raw/kc1.arff')

# Convert to a pandas DataFrame
df = pd.DataFrame(data[0])

print(df.head())

    loc  v(g)  ev(g)  iv(g)      n       v     l      d      i         e  ...  \
0   1.1   1.4    1.4    1.4    1.3    1.30  1.30   1.30   1.30      1.30  ...   
1   1.0   1.0    1.0    1.0    1.0    1.00  1.00   1.00   1.00      1.00  ...   
2  83.0  11.0    1.0   11.0  171.0  927.89  0.04  23.04  40.27  21378.61  ...   
3  46.0   8.0    6.0    8.0  141.0  769.78  0.07  14.86  51.81  11436.73  ...   
4  25.0   3.0    1.0    3.0   58.0  254.75  0.11   9.35  27.25   2381.95  ...   

   lOCode  lOComment  lOBlank  locCodeAndComment  uniq_Op  uniq_Opnd  \
0     2.0        2.0      2.0                2.0      1.2        1.2   
1     1.0        1.0      1.0                1.0      1.0        1.0   
2    65.0       10.0      6.0                0.0     18.0       25.0   
3    37.0        2.0      5.0                0.0     16.0       28.0   
4    21.0        0.0      2.0                0.0     11.0       10.0   

   total_Op  total_Opnd  branchCount   defects  
0       1.2         1.2        

In [185]:
print("Shape:")
print(df.shape)

print("\nColumn names:")
print(df.columns.tolist())

print("\nData types:")
print(df.dtypes)

print("\nMissing values:")
print(df.isnull().sum())

Shape:
(2109, 22)

Column names:
['loc', 'v(g)', 'ev(g)', 'iv(g)', 'n', 'v', 'l', 'd', 'i', 'e', 'b', 't', 'lOCode', 'lOComment', 'lOBlank', 'locCodeAndComment', 'uniq_Op', 'uniq_Opnd', 'total_Op', 'total_Opnd', 'branchCount', 'defects']

Data types:
loc                  float64
v(g)                 float64
ev(g)                float64
iv(g)                float64
n                    float64
v                    float64
l                    float64
d                    float64
i                    float64
e                    float64
b                    float64
t                    float64
lOCode               float64
lOComment            float64
lOBlank              float64
locCodeAndComment    float64
uniq_Op              float64
uniq_Opnd            float64
total_Op             float64
total_Opnd           float64
branchCount          float64
defects               object
dtype: object

Missing values:
loc                  0
v(g)                 0
ev(g)                0
iv(g)      

In [186]:
df["defects"].value_counts()

defects
b'false'    1783
b'true'      326
Name: count, dtype: int64

In [187]:
df["defects"].value_counts(normalize=True) * 100

defects
b'false'    84.542437
b'true'     15.457563
Name: proportion, dtype: float64

In [188]:
df.describe()

,loc,v(g),ev(g),iv(g),n,v,l,d,i,e,...,t,lOCode,lOComment,lOBlank,locCodeAndComment,uniq_Op,uniq_Opnd,total_Op,total_Opnd,branchCount
count,2109.000000,2109.000000,2109.000000,2109.000000,2109.000000,2109.000000,2109.000000,2109.000000,2109.000000,2109.000000,...,2109.00000,2109.000000,2109.000000,2109.000000,2109.000000,2109.000000,2109.000000,2109.000000,2109.000000,2109.000000
mean,20.372262,2.838028,1.674443,2.546420,49.829445,258.696719,0.319583,6.771242,21.240071,5242.386240,...,291.24504,14.525367,0.945946,1.759602,0.132764,7.631674,9.537316,31.043717,18.786724,4.665908
std,29.754442,3.900763,2.200659,3.375859,83.599874,516.317605,0.317029,7.863646,21.500367,17444.981211,...,969.16516,24.188302,3.085271,3.856850,0.704023,5.730347,12.195727,51.776056,32.074398,7.792206
min,1.000000,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.00000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000
25%,3.000000,1.000000,1.000000,1.000000,4.000000,8.000000,0.080000,1.500000,5.330000,12.000000,...,0.67000,0.000000,0.000000,0.000000,0.000000,3.000000,1.000000,3.000000,1.000000,1.000000
50%,9.000000,1.000000,1.000000,1.000000,16.000000,57.060000,0.200000,3.500000,14.400000,213.970000,...,11.89000,5.000000,0.000000,0.000000,0.000000,6.000000,5.000000,10.000000,6.000000,1.000000
75%,24.000000,3.000000,1.000000,3.000000,58.000000,265.930000,0.670000,9.200000,29.850000,2276.020000,...,126.45000,17.000000,0.000000,2.000000,0.000000,11.000000,13.000000,36.000000,22.000000,5.000000
max,288.000000,45.000000,26.000000,45.000000,1106.000000,7918.820000,2.000000,53.750000,193.060000,324803.510000,...,18044.64000,262.000000,44.000000,58.000000,12.000000,37.000000,120.000000,678.000000,428.000000,89.000000


## Initial EDA Findings
* The class distribution of the data set is imbalanced.
* The KC1 dataset consists of 2109 rows and 22 columns. Rows contain information on software modules and columns contain attributes of teh respective software modules.
* There are 22 attribute columns out of which 21 columns are used as explanatory variates (input of the model) and 1 is response variate (output of the model).
* There are no missing values found in the dataset.
* There are total of 1783 non-defective modules and 326 defective modules out of 2109 software modules.
* Several explanatory variables are rightly skewed because their means are larger than their medians.

## Day 2

In [189]:
X = df.drop(columns=["defects"])
Y = df["defects"]

print(X.shape)
print(Y.shape)

(2109, 21)
(2109,)


In [190]:
from sklearn.model_selection import train_test_split

In [191]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42, stratify=Y)

In [192]:
print(X_train.shape)
print(X_test.shape)
print(Y_train.shape)
print(Y_test.shape)

(1687, 21)
(422, 21)
(1687,)
(422,)


In [193]:
Y.value_counts(normalize=True)
Y_train.value_counts(normalize=True)
Y_test.value_counts(normalize=True)

defects
b'false'    0.845972
b'true'     0.154028
Name: proportion, dtype: float64

## Day 3

In [194]:
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, f1_score, roc_auc_score
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import make_pipeline

In [195]:
Y_train = Y_train.astype(str)
Y_test = Y_test.astype(str)

dummy = DummyClassifier(strategy="most_frequent")

In [196]:
dummy.fit(X_train, Y_train)
dummy_prediction = dummy.predict(X_test)

print(dummy_prediction[:20])

['false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false']


In [197]:
dummyAccuracy = accuracy_score(Y_test, dummy_prediction)
dummyConfusionMatrix = confusion_matrix(Y_test, dummy_prediction)

print("Accuracy:", dummyAccuracy)
print("Confusion Matrix:")
print(dummyConfusionMatrix)

Accuracy: 0.8459715639810427
Confusion Matrix:
[[357   0]
 [ 65   0]]


In [198]:
precision = precision_score(Y_test, dummy_prediction, pos_label="true", zero_division=0)
recall = recall_score(Y_test, dummy_prediction, pos_label="true")
f1 = f1_score(Y_test, dummy_prediction, pos_label="true", zero_division=0)

print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)

Precision: 0.0
Recall: 0.0
F1-score: 0.0


In [199]:
pipe = make_pipeline(StandardScaler(), LogisticRegression())

In [200]:
pipe.fit(X_train, Y_train)
pipe_prediction = pipe.predict(X_test)
pipeProbability = pipe.predict_proba(X_test)
pipe_defectiveProbability = pipeProbability[:, 1]

print(pipe_prediction[:20])

print("\nClasses:")
print(pipe.classes_)

print("\nProbability shape:")
print(pipeProbability.shape)

print("\nFirst 5 probability rows:")
print(pipeProbability[:5])

['false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false']

Classes:
['false' 'true']

Probability shape:
(422, 2)

First 5 probability rows:
[[0.96181716 0.03818284]
 [0.96539873 0.03460127]
 [0.84600489 0.15399511]
 [0.90963651 0.09036349]
 [0.86201079 0.13798921]]


In [201]:
pipeAccuracyScore = accuracy_score(Y_test, pipe_prediction)
pipeConfusionMatrix = confusion_matrix(Y_test, pipe_prediction)
pipePrecision = precision_score(Y_test, pipe_prediction, pos_label="true", zero_division=0)
pipeRecall = recall_score(Y_test, pipe_prediction, pos_label="true")
pipeF1 = f1_score(Y_test, pipe_prediction, pos_label="true", zero_division=0)
pipeROC_AUC = roc_auc_score(Y_test, pipe_defectiveProbability)

print("Accuracy: ", pipeAccuracyScore)
print("\nConfusion Matrix: ")
print(pipeConfusionMatrix)

print("\nPrecision:", pipePrecision)
print("\nRecall:", pipeRecall)
print("\nF1-score:", pipeF1)
print("\nROC-AUC Score:", pipeROC_AUC)

Accuracy:  0.8601895734597157

Confusion Matrix: 
[[349   8]
 [ 51  14]]

Precision: 0.6363636363636364

Recall: 0.2153846153846154

F1-score: 0.3218390804597701

ROC-AUC Score: 0.8040077569489334


## Day 3 Findings

The DummyClassifier baseline achieved about 84.6% accuracy but detected 0 defective modules. Logistic Regression achieved about 86.0% accuracy and detected 14 defective modules, but it still missed 51 of the 65 defective modules in the test set. This shows that Logistic Regression improves over the baseline, but it still has difficulty identifying defective modules. Therefore, accuracy alone does not give a complete picture of the model's performance, especially because the dataset is imbalanced.

## Day 4

In [202]:
from sklearn.ensemble import RandomForestClassifier

In [203]:
randomForest = RandomForestClassifier(random_state=42)

In [204]:
randomForest.fit(X_train, Y_train)

randomForest_prediction = randomForest.predict(X_test)
randomForest_probability = randomForest.predict_proba(X_test)
randomForest_defectiveProbability =randomForest_probability[:, 1]

print(randomForest_prediction[:20])

print("\nClasses:")
print(randomForest.classes_)

print("\nProbability shape:")
print(randomForest_probability.shape)

print("\nFirst 5 probability rows:")
print(randomForest_probability[:5])


['false' 'false' 'true' 'false' 'false' 'false' 'false' 'false' 'true'
 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false']

Classes:
['false' 'true']

Probability shape:
(422, 2)

First 5 probability rows:
[[1.     0.    ]
 [1.     0.    ]
 [0.27   0.73  ]
 [0.8975 0.1025]
 [0.97   0.03  ]]


In [205]:
randomForest_confusionMatrix = confusion_matrix(Y_test, randomForest_prediction)
randomForest_accuracy = accuracy_score(Y_test, randomForest_prediction)
randomForest_precision = precision_score(Y_test, randomForest_prediction, pos_label="true", zero_division=0)
randomForest_recall = recall_score(Y_test, randomForest_prediction, pos_label="true", zero_division=0)
randomForest_F1 = f1_score(Y_test, randomForest_prediction, pos_label="true", zero_division=0)
randomForest_ROC_AUC = roc_auc_score(Y_test, randomForest_defectiveProbability)

print("Confusion Matrix: ")
print(randomForest_confusionMatrix)
print("\nAccuracy: ", randomForest_accuracy)
print("\nPrecision:", randomForest_precision)
print("\nRecall:", randomForest_recall)
print("\nF1-score:", randomForest_F1)
print("\nROC-AUC Score:", randomForest_ROC_AUC)

Confusion Matrix: 
[[338  19]
 [ 40  25]]

Accuracy:  0.8601895734597157

Precision: 0.5681818181818182

Recall: 0.38461538461538464

F1-score: 0.45871559633027525

ROC-AUC Score: 0.8350570997629821


## Day 4 Findings

Logistic Regression and Random Forest both achieved about 86.0% accuracy on the same KC1 test set. However, their defect-detection performance was different. Random Forest detected 25 of the 65 defective modules, while Logistic Regression detected only 14. Therefore, Random Forest had higher recall, F1-score, and ROC-AUC, while Logistic Regression had higher precision.

The equal accuracy does not mean that the two models performed equally well for defect detection. Since the KC1 dataset is imbalanced, accuracy can remain high even when a model misses many defective modules. Random Forest was better at identifying defective modules under this evaluation. This shows why accuracy alone is not sufficient for evaluating software defect prediction models.

## Day 5

In [206]:
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import cross_validate
from sklearn.metrics import make_scorer, precision_score, recall_score, f1_score

In [207]:
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

In [208]:
metric_names = {
    "accuracy": "accuracy",
    "precision": make_scorer(precision_score, pos_label="true"),
    "recall": make_scorer(recall_score, pos_label="true"),
    "f1": make_scorer(f1_score, pos_label="true"),
    "roc_auc": "roc_auc"
}

In [209]:
baseline_cv = cross_validate(estimator=dummy, X=X_train, y=Y_train, cv=skf, scoring=metric_names)
print(baseline_cv, "\n")

{'fit_time': array([0.00118494, 0.00087881, 0.00106907, 0.00214386, 0.00106788]), 'score_time': array([0.01032615, 0.0102911 , 0.00924611, 0.01016021, 0.01026106]), 'test_accuracy': array([0.84615385, 0.84319527, 0.84569733, 0.84569733, 0.84569733]), 'test_precision': array([0., 0., 0., 0., 0.]), 'test_recall': array([0., 0., 0., 0., 0.]), 'test_f1': array([0., 0., 0., 0., 0.]), 'test_roc_auc': array([0.5, 0.5, 0.5, 0.5, 0.5])} 



/Users/saiyamrajput/Desktop/Work/Software Defect Prediction/.venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/Users/saiyamrajput/Desktop/Work/Software Defect Prediction/.venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/Users/saiyamrajput/Desktop/Work/Software Defect Prediction/.venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter

In [210]:
print("Accuracy mean:", baseline_cv["test_accuracy"].mean())
print("Accuracy std:", baseline_cv["test_accuracy"].std())

print("\nPrecision mean:", baseline_cv["test_precision"].mean())
print("Precision std:", baseline_cv["test_precision"].std())

print("\nRecall mean:", baseline_cv["test_recall"].mean())
print("Recall std:", baseline_cv["test_recall"].std())

print("\nF1 mean:", baseline_cv["test_f1"].mean())
print("F1 std:", baseline_cv["test_f1"].std())

print("\nROC-AUC mean:", baseline_cv["test_roc_auc"].mean())
print("ROC-AUC std:", baseline_cv["test_roc_auc"].std())

Accuracy mean: 0.8452882201113198
Accuracy std: 0.0010613081919659422

Precision mean: 0.0
Precision std: 0.0

Recall mean: 0.0
Recall std: 0.0

F1 mean: 0.0
F1 std: 0.0

ROC-AUC mean: 0.5
ROC-AUC std: 0.0


In [211]:
logistic_cv = cross_validate(estimator=pipe, X=X_train, y=Y_train, cv=skf, scoring=metric_names)
print(logistic_cv, "\n")

{'fit_time': array([0.01099396, 0.01423192, 0.00956392, 0.01524234, 0.01359415]), 'score_time': array([0.01353168, 0.01081514, 0.01126909, 0.01269364, 0.01496601]), 'test_accuracy': array([0.84911243, 0.84911243, 0.85163205, 0.87240356, 0.86053412]), 'test_precision': array([0.52380952, 0.5625    , 0.5625    , 0.73684211, 0.69230769]), 'test_recall': array([0.21153846, 0.16981132, 0.17307692, 0.26923077, 0.17307692]), 'test_f1': array([0.30136986, 0.26086957, 0.26470588, 0.3943662 , 0.27692308]), 'test_roc_auc': array([0.80796127, 0.7795432 , 0.78704453, 0.84908907, 0.8268556 ])} 



In [212]:
print("Accuracy mean:", logistic_cv["test_accuracy"].mean())
print("Accuracy std:", logistic_cv["test_accuracy"].std())

print("\nPrecision mean:", logistic_cv["test_precision"].mean())
print("Precision std:", logistic_cv["test_precision"].std())

print("\nRecall mean:", logistic_cv["test_recall"].mean())
print("Recall std:", logistic_cv["test_recall"].std())

print("\nF1 mean:", logistic_cv["test_f1"].mean())
print("F1 std:", logistic_cv["test_f1"].std())

print("\nROC-AUC mean:", logistic_cv["test_roc_auc"].mean())
print("ROC-AUC std:", logistic_cv["test_roc_auc"].std())

Accuracy mean: 0.8565589170017383
Accuracy std: 0.008966933840272532

Precision mean: 0.6155918642760747
Precision std: 0.08324473477180236

Recall mean: 0.19934687953555877
Recall std: 0.0381705728745054

F1 mean: 0.2996469169380413
F1 std: 0.04942801311542105

ROC-AUC mean: 0.8100987341790183
ROC-AUC std: 0.025576334619452697


In [213]:
randomForest_cv = cross_validate(estimator=randomForest, X=X_train, y=Y_train, cv=skf, scoring=metric_names)
print(randomForest_cv, "\n")

{'fit_time': array([0.21718192, 0.21051693, 0.13979697, 0.16254306, 0.15415096]), 'score_time': array([0.02431393, 0.01733828, 0.02114582, 0.02211595, 0.02004695]), 'test_accuracy': array([0.86390533, 0.85207101, 0.85459941, 0.86350148, 0.86350148]), 'test_precision': array([0.625     , 0.54545455, 0.55555556, 0.625     , 0.65      ]), 'test_recall': array([0.28846154, 0.33962264, 0.28846154, 0.28846154, 0.25      ]), 'test_f1': array([0.39473684, 0.41860465, 0.37974684, 0.39473684, 0.36111111]), 'test_roc_auc': array([0.84080823, 0.80923535, 0.80522942, 0.81349528, 0.81700405])} 



In [214]:
print("Accuracy mean:", randomForest_cv["test_accuracy"].mean())
print("Accuracy std:", randomForest_cv["test_accuracy"].std())

print("\nPrecision mean:", randomForest_cv["test_precision"].mean())
print("Precision std:", randomForest_cv["test_precision"].std())

print("\nRecall mean:", randomForest_cv["test_recall"].mean())
print("Recall std:", randomForest_cv["test_recall"].std())

print("\nF1 mean:", randomForest_cv["test_f1"].mean())
print("F1 std:", randomForest_cv["test_f1"].std())

print("\nROC-AUC mean:", randomForest_cv["test_roc_auc"].mean())
print("ROC-AUC std:", randomForest_cv["test_roc_auc"].std())

Accuracy mean: 0.8595157410496373
Accuracy std: 0.005111460967204851

Precision mean: 0.6002020202020202
Precision std: 0.041714053250038984

Recall mean: 0.29100145137880984
Recall std: 0.028511375336062524

F1 mean: 0.3897872563854932
F1 std: 0.018987208324787903

ROC-AUC mean: 0.8171544655405706
ROC-AUC std: 0.012472625559066357


In [215]:
models = ["DummyClassifier", "Logistic Regression", "Random Forest Classifier"]
meanTableAccuracy = [baseline_cv["test_accuracy"].mean(), logistic_cv["test_accuracy"].mean(), randomForest_cv["test_accuracy"].mean()]
meanTablePrecision = [baseline_cv["test_precision"].mean(), logistic_cv["test_precision"].mean(), randomForest_cv["test_precision"].mean()]
meanTableRecall = [baseline_cv["test_recall"].mean(), logistic_cv["test_recall"].mean(), randomForest_cv["test_recall"].mean()]
meanTableF1 = [baseline_cv["test_f1"].mean(), logistic_cv["test_f1"].mean(), randomForest_cv["test_f1"].mean()]
meanTableRocAuc = [baseline_cv["test_roc_auc"].mean(), logistic_cv["test_roc_auc"].mean(), randomForest_cv["test_roc_auc"].mean()]

meanComparisonTable = pd.DataFrame({
    "Model": models,
    "Accuracy Mean": meanTableAccuracy,
    "Precision Mean": meanTablePrecision,
    "Recall Mean": meanTableRecall,
    "F1 Mean": meanTableF1,
    "ROC-AUC Mean": meanTableRocAuc
})

print(meanComparisonTable)

                      Model  Accuracy Mean  Precision Mean  Recall Mean  \
0           DummyClassifier       0.845288        0.000000     0.000000   
1       Logistic Regression       0.856559        0.615592     0.199347   
2  Random Forest Classifier       0.859516        0.600202     0.291001   

    F1 Mean  ROC-AUC Mean  
0  0.000000      0.500000  
1  0.299647      0.810099  
2  0.389787      0.817154  


In [216]:
stdTableAccuracy = [baseline_cv["test_accuracy"].std(), logistic_cv["test_accuracy"].std(), randomForest_cv["test_accuracy"].std()]
stdTablePrecision = [baseline_cv["test_precision"].std(), logistic_cv["test_precision"].std(), randomForest_cv["test_precision"].std()]
stdTableRecall = [baseline_cv["test_recall"].std(), logistic_cv["test_recall"].std(), randomForest_cv["test_recall"].std()]
stdTableF1 = [baseline_cv["test_f1"].std(), logistic_cv["test_f1"].std(), randomForest_cv["test_f1"].std()]
stdTableRocAuc = [baseline_cv["test_roc_auc"].std(), logistic_cv["test_roc_auc"].std(), randomForest_cv["test_roc_auc"].std()]

stdComparisonTable = pd.DataFrame({
    "Model STD": models,
    "Accuracy STD": stdTableAccuracy,
    "Precision STD": stdTablePrecision,
    "Recall STD": stdTableRecall,
    "F1 STD": stdTableF1,
    "ROC-AUC STD": stdTableRocAuc
})

print(stdComparisonTable)

                  Model STD  Accuracy STD  Precision STD  Recall STD  \
0           DummyClassifier      0.001061       0.000000    0.000000   
1       Logistic Regression      0.008967       0.083245    0.038171   
2  Random Forest Classifier      0.005111       0.041714    0.028511   

     F1 STD  ROC-AUC STD  
0  0.000000     0.000000  
1  0.049428     0.025576  
2  0.018987     0.012473  


In [217]:
finalComparisonTable = pd.DataFrame({
    "Model": models,
    "Accuracy": [f"{meanTableAccuracy[i]:.3f} ± {stdTableAccuracy[i]:.3f}" for i in range(3)],
    "Precision": [f"{meanTablePrecision[i]:.3f} ± {stdTablePrecision[i]:.3f}" for i in range(3)],
    "Recall": [f"{meanTableRecall[i]:.3f} ± {stdTableRecall[i]:.3f}" for i in range(3) ],
    "F1": [f"{meanTableF1[i]:.3f} ± {stdTableF1[i]:.3f}" for i in range(3)],
    "ROC-AUC": [f"{meanTableRocAuc[i]:.3f} ± {stdTableRocAuc[i]:.3f}" for i in range(3)]
})

print(finalComparisonTable)

                      Model       Accuracy      Precision         Recall  \
0           DummyClassifier  0.845 ± 0.001  0.000 ± 0.000  0.000 ± 0.000   
1       Logistic Regression  0.857 ± 0.009  0.616 ± 0.083  0.199 ± 0.038   
2  Random Forest Classifier  0.860 ± 0.005  0.600 ± 0.042  0.291 ± 0.029   

              F1        ROC-AUC  
0  0.000 ± 0.000  0.500 ± 0.000  
1  0.300 ± 0.049  0.810 ± 0.026  
2  0.390 ± 0.019  0.817 ± 0.012  


## Day 5 Findings

The DummyClassifier achieved about 84.5% accuracy, but it had zero precision, recall, and F1 with a ROC-AUC of 0.5. This shows that the baseline can achieve high accuracy by predicting the majority class while failing to detect defective modules. Logistic Regression clearly outperformed the DummyClassifier across the defect-detection metrics.
Random Forest had higher recall (0.291), F1-score (0.390), and ROC-AUC (0.817) than Logistic Regression, while Logistic Regression had slightly higher precision (0.616). Therefore, Random Forest was stronger overall for defect detection under the 5-fold cross-validation evaluation.
However, Random Forest's mean recall was only about 29%, meaning it still missed many defective modules. This shows that although Random Forest performed better than Logistic Regression, there is still significant room for improvement in detecting defective software modules.
